client-apis/encrypt.md

```diff
@@ -113,68 +161,168 @@ This document refers to the output of the key derivation algorithm as the derive
+  as determined above.
+- [IV](../data-format/message-body.md#iv): MUST be the IV used when calculating the encrypted content above
+- [Encrypted Content](../data-format/message-body.md#encrypted-content): MUST be the encrypted content calculated above.
+- [Authentication Tag](../data-format/message-body.md#authentication-tag): MUST be the authentication tag
+  outputted when calculating the encrypted content above.
+
+If the algorithm suite contains a signature algorithm and
+this operation is streaming the encrypted message output to the caller,
+this operation MUST input the serialized frame to the signature algorithm as soon as it is serialized,
+such that the serialized frame isn't required to remain in memory to complete
+the [signature calculation](#signature-calculation).
+
+### Construct the Signature
 
 If the [algorithm suite](../framework/algorithm-suites.md) contains a [signature algorithm](../framework/algorithm-suites.md#signature-algorithm),
-the output [message](../data-format/message.md) MUST contain a [message footer](../data-format/message-footer.md).
-The footer is serialized with the following specifics:
+this operation MUST calculate a signature over the message,
+and the output [encrypted message](#encrypted-message) MUST contain a [message footer](../data-format/message-footer.md).
+
+To calculate a signature, this operation MUST use the [signature algorithm](../framework/algorithm-suites.md#signature-algorithm)
+specified by the [algorithm suite](../framework/algorithm-suites.md), with the following input:
+
+- the signature key is the [signing key](../framework/structures.md#signing-key) in the [encryption materials](../framework/structures.md#encryption-materials)
+- the input to sign is the concatenation of the serialization of the [message header](../data-format/message-header.md) and [message body](../data-format/message-body.md)
 
-- [Signature](../data-format/message-footer.md#signature): MUST be the output of the [signature algorithm](../framework/algorithm-suites.md#signature-algorithm)
-  specified by the [algorithm suite](../framework/algorithm-suites.md), with the following input:
-  - the signature key is the [signing key](../framework/structures.md#signing-key) in the [encryption materials](../framework/structures.md#encryption-materials)
-  - the input to sign is the concatenation of the serialization of the [message header](../data-format/message-header.md) and [message body](../data-format/message-body.md)
+Note that the message header and message body MAY have already been inputted during previous steps.
 
-Any data that is not specified within the [message format](../data-format/message.md) MUST NOT be added to the output message.
+This operation MUST then serialize serialized a message footer with the following specifics:
+
+- [Signature](../data-format/message-footer.md#signature): MUST be the output of the calculation above.
```

Comment:
Well, not exactly.  Since the above calculation does not include a length.

Reply:
Added spec for the signature length field.